[DAG]Add ISD::SPLAT_VECTOR to TargetLowering::getNegatedExpression

ChunyuLiao opened the pull request:
Fold splat_vector(fneg(X)) -> splat_vector(-X)
Call the getCheaperNegatedExpression function, and ISD::SPLAT_VECTOR return NegatibleCost::Cheaper. 
This optimization is applied only to the fneg instruction.

ChunyuLiao committed:
init testcase

ChunyuLiao committed:
[DAGCombine]Fold (fadd A, Splat(fneg(B))) -> (fsub A, Splat(B))

RKSimon reviewed (commented):
Can the getNegatedExpression code not handle splats?

ChunyuLiao committed:
Revert "[DAGCombine]Fold (fadd A, Splat(fneg(B))) -> (fsub A, Splat(B))"

This reverts commit 53feacfeb2b638ce1a60bde313a52b8cadc4a283.

ChunyuLiao committed:
[DAG]Add splat_vector to TargetLowering::getNegatedExpression
splat_vector(fneg(x)) to splat_vector(x),
then DAGCombine performs some optimizations using these results.

ChunyuLiao committed:
Revert "[DAG]Add splat_vector to TargetLowering::getNegatedExpression"

This reverts commit 30f4c8f230eee7b3d76b43af781bfdf1b3d2454c.

ChunyuLiao committed:
[DAGCombine]Fold (fadd A, Splat(fneg(B))) -> (fsub A, Splat(B))
At the same time,fold (fsub A, Splat(fneg(B))) -> (fadd A, Splat(B))

ChunyuLiao committed:
Add TLI.isFNegFree(VT)

ChunyuLiao commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17965,6 +17965,19 @@ SDValue DAGCombiner::visitFADD(SDNode *N) {
+      (!LegalOperations || TLI.isOperationLegalOrCustom(ISD::FSUB, VT))) {
+    if (N1.getOpcode() == ISD::SPLAT_VECTOR) {
+      SDValue SplatN0 = N1->getOperand(0);
+      if (SplatN0.getOpcode() == ISD::FNEG && SplatN0.hasOneUse()) {
```
In the last commit, I added a check for TLI.isFNegFree(VT) without modifying any test cases. If  remove hasOneUse(), there will be no optimization effect for RISCV for now.

ChunyuLiao committed:
Revert "Add TLI.isFNegFree(VT)"

This reverts commit 5649f0ae622474142b9ed13f0df09b374b4a5a81.

ChunyuLiao committed:
Revert "[DAGCombine]Fold (fadd A, Splat(fneg(B))) -> (fsub A, Splat(B))"

This reverts commit 4937d222e08a7a7a604920b7a443300be47a461f.

ChunyuLiao committed:
[DAGCombine]Fold (fadd A, Splat(fneg(B))) -> (fsub A, Splat(B))
Fold (fsub A, Splat(fneg(B))) -> (fadd A, Splat(B))

ChunyuLiao commented:
> @ChunyuLiao Sorry I don't quite understand, why don't you want to use the existing getCheaperNegatedExpression call? I don't like the idea that we end up with duplicated folds for splat cases everywhere.

Following this suggestion, I tried again, and this time it worked. I might have made a mistake somewhere before, which caused the infinite loop. Anyway, thank you so much!

ChunyuLiao committed:
Add todo for ISD::SPLAT_VECTOR

ChunyuLiao committed:
[DAG]Add ISD::SPLAT_VECTOR to TargetLowering::getNegatedExpression
Fold splat_vector(fneg(X)) -> splat_vector(-X)

RKSimon reviewed (approved):
LGTM - cheers!
